### PR TITLE
Replacing chkconfig commnand with systemctl

### PR DESCRIPTION
**Pull request by hegerj**

### CODE REVIEW

**Comment by hegerj** [comment]
There are still functions called `chkconfig()` in test/infrastructureTest.sh but I am not sure how that works because it doesn't seem that they are called...? Anyway they do not call `chkconfig` command itself so it should not be a problem.

**Comment by sopos** [comment]
> There are still functions called `chkconfig()` in test/infrastructureTest.sh but I am not sure how that works because it doesn't seem that they are called...? Anyway they do not call `chkconfig` command itself so it should not be a problem.

Those should be just a fake funftions for the testing purposes (not to mess with the system).

**Comment by sopos** [comment]
I'm a bit afraid just replacing it might not be a good idea as it may not work correctly with older rhels. Namely rhel6 do not use `systemctl` at all and we recently needed to created rhel6 build.

I would prefer similar solution as [__INTERNAL_SERVICE](https://github.com/beakerlib/beakerlib/blob/master/src/infrastructure.sh#L969) with a [fallback](https://github.com/beakerlib/beakerlib/blob/master/src/infrastructure.sh#L979) if the want command does not exist.

Such approach would prefer the old way if available. This may prevent similar situations as `service auditd stop` vs. `systemctl stop auditd` where systemctl does not correctly.

File: src/infrastructure.sh [hunk 1]
```diff
@@ -1372,7 +1372,7 @@ rlServiceEnable() {
```
Comment (sopos) [inline src/infrastructure.sh:1372]: something like
```suggestion
# __INTERNAL_CHKCONFIG operation..
# a translation layer from systemctl-like call to chkconfig
# note that the parameters order is matching systemctl
# $1 - operation
# $2 - service
__INTERNAL_CHKCONFIG() {
  local svc="$2" op
  case $1 in
    is-enabled)
      op=""
    ;;
    enable)
      op="on"
    ;;
    disable)
      op="off"
    ;;
    *)
      rlLogError "chkconfig: operation not implement, please cover!"
    ;;
  esac
  chkconfig "$svc" $op
}
which chkconfig >/dev/null 2>&1 || __INTERNAL_CHKCONFIG() { __INTERNAL_SYSTEMCTL "$@"; }
```

File: src/infrastructure.sh [hunk 2]
```diff
@@ -1372,7 +1372,7 @@ rlServiceEnable() {
 
   local service
   for service in "$@"; do
-      chkconfig $service > /dev/null
+      systemctl is-enabled $service > /dev/null
```
Comment (sopos) [inline src/infrastructure.sh:1375]:
```suggestion
      __INTERNAL_CHKCONFIG is-enabled $service > /dev/null
```

File: src/infrastructure.sh [hunk 3]
```diff
@@ -1385,7 +1385,7 @@ rlServiceEnable() {
       fi
 
       # enable service
-      if chkconfig "$service" on; then
+      if systemctl enable "$service"; then
```
Comment (sopos) [inline src/infrastructure.sh:1388]:
```suggestion
      if __INTERNAL_CHKCONFIG enable "$service"; then
```

File: src/infrastructure.sh [hunk 4]
```diff
@@ -1439,7 +1439,7 @@ rlServiceDisable() {
 
     local service
     for service in "$@"; do
-        chkconfig $service > /dev/null
+        systemctl is-enabled $service > /dev/null
```
Comment (sopos) [inline src/infrastructure.sh:1442]:
```suggestion
        __INTERNAL_CHKCONFIG is-enabled $service > /dev/null
```

File: src/infrastructure.sh [hunk 5]
```diff
@@ -1452,7 +1452,7 @@ rlServiceDisable() {
         fi
 
         # disable it
-        if chkconfig "$service" off; then
+        if systemctl disable "$service"; then
```
Comment (sopos) [inline src/infrastructure.sh:1455]:
```suggestion
        if __INTERNAL_CHKCONFIG disable "$service"; then
```

File: src/infrastructure.sh [hunk 6]
```diff
@@ -1564,24 +1564,24 @@ __INTERNAL_SOCKET_service() {
   rlLogDebug "rlSocket: Handling $serviceName via xinetd"
   case $serviceTask in
     "start")
-      rlServiceStart xinetd && chkconfig ${serviceName} on
+      rlServiceStart xinetd && systemctl enable ${serviceName}
```
Comment (sopos) [inline src/infrastructure.sh:1567]:
```suggestion
      rlServiceStart xinetd && __INTERNAL_CHKCONFIG enable ${serviceName}
```

File: src/infrastructure.sh [hunk 7]
```diff
@@ -1564,24 +1564,24 @@ __INTERNAL_SOCKET_service() {
...
       fi
     ;;
     "stop")
-      chkconfig ${serviceName} off
+      systemctl disable ${serviceName}
       return
     ;;
     "status")
-      chkconfig ${serviceName} > /dev/null;   local outcome=$?
+      systemctl is-enabled ${serviceName} > /dev/null;   local outcome=$?
```
Comment (sopos) [inline src/infrastructure.sh:1584]:
```suggestion
      __INTERNAL_CHKCONFIG is-enabled ${serviceName} > /dev/null;   local outcome=$?
```

File: src/infrastructure.sh [hunk 8]
```diff
@@ -1564,24 +1564,24 @@ __INTERNAL_SOCKET_service() {
...
+        systemctl is-enabled ${serviceName} > /dev/null;   local outcome=$?
         __INTERNAL_SERVICE status xinetd 2>&1 > /dev/null; local outcomeXinetd=$?
         rlLogDebug "xinetd status code: $outcomeXinetd"
         rlLogDebug "socket $serviceName status: $outcome"
         return 1
       fi
     ;;
     "stop")
-      chkconfig ${serviceName} off
+      systemctl disable ${serviceName}
```
Comment (sopos) [inline src/infrastructure.sh:1580]:
```suggestion
      __INTERNAL_CHKCONFIG disable ${serviceName}
```

File: src/infrastructure.sh [hunk 9]
```diff
@@ -1564,24 +1564,24 @@ __INTERNAL_SOCKET_service() {
...
   case $serviceTask in
     "start")
-      rlServiceStart xinetd && chkconfig ${serviceName} on
+      rlServiceStart xinetd && systemctl enable ${serviceName}
       outcome=$?
       if [[ $outcome == "0" ]]; then
         return 0
       else
-        chkconfig ${serviceName} > /dev/null;   local outcome=$?
+        systemctl is-enabled ${serviceName} > /dev/null;   local outcome=$?
```
Comment (sopos) [inline src/infrastructure.sh:1572]:
```suggestion
        __INTERNAL_CHKCONFIG is-enabled ${serviceName} > /dev/null;   local outcome=$?
```

**Review by sopos — CHANGES_REQUESTED** [review]

**Comment by hegerj** [comment]
I realise the RHEL6 doesn't have systemd however I thought it would be better to patch the problem on potential next RHEL6 release rather then do full release. It might be bit more work then rebase but at the cost of adding another 'wrapper' function as you suggested that, in my opinion, makes the code quite hard to understand (constant jumping from function to function, overriding and nested functions makes it harder to find issues/read the code).
But if you think this solution is better we can do it. 
Good catch about `__INTERNAL_SYSTEMCTL` which I should have used.

**Comment by sopos** [comment]
The thing is that now we have a common spec file so the doing a rebase in downstream is easier than adding patches. But we can crate the patch even in the upstream and distribute it along way.

**Comment by sopos** [comment]
I'm not really sure which which approach would be better. In case of RHEL, there's `chkconfig` available in RHEL-8 as well. So we might be all but Fedora?

**Comment by hegerj** [comment]
> The thing is that now we have a common spec file so doing a rebase in downstream is easier than adding patches. But we can create the patch even in the upstream and distribute it along the way. However I consider maintaining a patch more demanding.

ok, that makes sense, no patching then

> I'm not really sure which which approach would be better. In case of RHEL, there's chkconfig available in RHEL-8 as well. So we might be all but Fedora?

Not sure what you mean but `chkconfig` is available on all Fedoras (31,32,33) it's just not installed by default I guess.

**Comment by sopos** [comment]
> Not sure what you mean but `chkconfig` is available on all Fedoras (31,32,33) it's just not installed by default I guess.

fixed the text in the original comment

`chkconfi`g may be installed on Fedora but I'm talking about the fact that `chkconfig` is installed by _default_ in some distro versions. Apparently it was by default installed on Fedora upto 33 but currently Fedora 34 does not.

**Comment by sopos** [comment]
I've provided a commit implementing the falback mechanism, its main advantage is virtually no additional action in case of downstream build since rhel6.

**Review by sopos — APPROVED** [review]

**Pull request merged**